src/transport/manager/mod.rs

```diff
@@ -381,16 +381,16 @@ impl TransportManager {
         self.transport_manager_handle.register_transport(name);
     }
 
+    /// Get the list of public addresses of the node.
+    pub(crate) fn public_addresses(&self) -> PublicAddresses {
+        self.listen_addresses.clone()
+    }
+
     /// Register local listen address.
     pub fn register_listen_address(&mut self, address: Multiaddr) {
-        assert!(!address.iter().any(|protocol| std::matches!(protocol, Protocol::P2p(_))));
-
-        let mut listen_addresses = self.listen_addresses.write();
-
-        listen_addresses.insert(address.clone());
-        listen_addresses.insert(address.with(Protocol::P2p(
-            Multihash::from_bytes(&self.local_peer_id.to_bytes()).unwrap(),
-        )));
+        if let Err(address) = self.listen_addresses.add_public_address(address) {
+            tracing::warn!(target: LOG_TARGET, ?address, "failed to register listen address");
```

Comment:
May be extend the error returned from `add_public_address()` to indicate a reason? Not super important, feel free to ignore.

Reply:
Yep I like that idea, have added a small error enum